### DISCUSSION TITLE
Should I create too many topics in Kafka?

### DISCUSSION THREAD

**Discussion by clyang82** [strimzi/strimzi-kafka-operator#9929] [General]
My requirement is as a cluster admin, I can isolate my cluster data so that the user can only access the data/message who has permission to do. In order to meet this requirement, the intuitional solution is to leverage the existing feature to define the kafka user to access the certain topics with pre-defined ACLs.

If I want to manage many clusters, for example 10K managed clusters, I need to create 10K kafkausers and 10K topics at least (one user and one topic per cluster). I searched if Kafka can support many topics. From [confluent document](https://developer.confluent.io/faq/apache-kafka/topics-in-kafka/#topics-in-kafka-how-to-create-a-topic-in-kafka)
> While there is no set limit to the number of topics that can exist in a Kafka cluster, currently Kafka can handle hundreds of thousands of topics, depending on the number of partitions in each.

It seems I can do in this way. So I use the admin API to create 10K users and 10K topics with ACLs ( It has problem to create 10K kafkaUser and kafkaTopic CR), it works. But I guess it may have the performance issues if the action triggers the re-balance. Does this solution make sense for you to support large scale with ACLs? any suggestions for large scale environment with authorization? Thanks in advance.

**Comment by scholzj** [comment]
>If I want to manage many clusters, for example 10K managed clusters, I need to create 10K kafkausers and 10K topics at least (one user and one topic per cluster).

I think it depends on what exactly are you asking for: 
* 10k Kafka clusters each with 10k topics and users each? How would that match against Kubernetes clusters? Would each be in its own Kubernetes clusters? Or will they share the Kubernetes cluster? If they would share the same Kube cluster, I would expect that you would run into some resource limits with Kubernetes itself.
* 10k Kafka clusters with 1 topic and 1 user each? Again, it would depend on how that would match the Kubernetes clusters. If it would be one Kubernetes cluster, 10k custom resources of each might be fine. But you will likely run into limitations of how many Kafka clusters the Strimzi Cluster Operator can handle. Also, this will cost you a huge amount of resources just to run the Kafka clusters. So I'm not sure what the purpose of this really is, but it does not sound like the right way to do it.

Of course, using the User and Topic operators is completely optional. So you can also manage things directly using Kafka Admin API if you want.

**Reply by clyang82** [reply]
Thanks @scholzj for your reply. Sorry for misleading. The cluster I mentioned is a Kubernetes cluster. It is not a Kafka Cluster. I draw this chart to clarify:
<img width="583" alt="image" src="https://github.com/strimzi/strimzi-kafka-operator/assets/7711311/29601e28-683f-4881-8dcd-e6b16855e5af">
1. there is only one kafka cluster provisioned by Strimzi Cluster Operator in the hub kubernetes cluster.
2. We want to manage 10K managed cluster by one hub cluster. One managed cluster will have one kafka topic and one kafka user so that we can do authorization by leveraging the ACLs feature.
so we want to have one kafka cluster to support 10K kafka topics and 10K kafka user. Does it make sense? Thanks.

**Reply by scholzj** [reply]
I guess I don't understand what this _managed cluster_ is. The User Operator is tested against 1000 users. I have no idea how well it handles 10000. You would need to try it how it handles 10000 users. @fvaleri, @tombentley or @see-quick might know more about what the Unidirectional Topic Operator was tested for.

**Reply by clyang82** [reply]
managed cluster is a Kubernetes cluster. We will have an agent in each managed cluster. The agent launches kafka producer or kafka consumer to publish or receive messages from the certain topic.

**Reply by scholzj** [reply]
So, from the overall architecture perspective -> what kind of data will it be sending / receiving? Kafka in general can handle 10000 topics if they have reasonable amount of partitions. But it makes one wonder if you have the right architecture. Once you have something like single topic per client etc. it starts to be suspicious that maybe you are not doing the right thing.

**Reply by fvaleri** [reply]
It's important to know how many partition do you have on these 10k topics, because this is the unit of replication in Kafka. I'm also questioning the use case behind this requirement.

What really matters for the Topic Operator is the rate of change events (create/update/delete). Are these 10k topics going to be created all at once? What about configuration changes? Are they managed centrally and batched together, or every app can change topic configuration at will?

If it helps, I was able to reconcile 10k change events in under 10 minutes with some tuning. That was on my laptop with Minikube, so I guess it can do even better on a real Kube cluster with enough resources. Also, keep in mind that it uses a batching system for Kafka admin operations that can be tuned.

**Reply by see-quick** [reply]
> @fvaleri, @tombentley or @see-quick might know more about what the Unidirectional Topic Operator was tested for.

So, I have tried to test the Unidirectional Topic Operator to create `5000 KafkaTopics`. It took 12 minutes to create those all at once but for that, you should have a solid Kubernetes cluster (which I assume you would have). It also really depends on the `KafkaTopic` configuration as @fvaleri pointed out...(i.e., replication factor + partition count)

For example, in my test scenario, I tried to create `KafkaTopics` with `replication factor=3` and `partitions count=12`. I had `9 broker nodes` where each node consumed `6GiB of memory` (with producers this number will be definitely higher...just to make a picture about numbers...). Most likely 10k topics => 10GiB memory?

I think at first we should understand the use case behind it (@fvaleri questions) and then we could find the optimal possible configuration for the Topic Operator (batch size, linger times etc.).

---
Now I have tried to deploy 10k KafkaTopics and It took around 40 minutes.

Memory footprint:
- TO container: `~900 MiB`
- for each Kafka pod: `~8 GiB` and `3-4 CPUs`

and some additional about configuration and results:
```
Number of Topics: 10000
STRIMZI_MAX_BATCH_SIZE: 2000
MAX_BATCH_LINGER_MS: 3000
Reconciliation Max Duration (s): 48.725
Max Batch Size: 2000.0
Max Queue Size: 180613.0
UTO Event Queue Time (s): 3093361.86
Describe Configs Max Duration (s): 2.053
Create Topics Max Duration (s): 29.543
Reconciliations Max Duration (s): 48.725
Update Status Max Duration (s): 0.243
```

**Reply by scholzj** [reply]
@see-quick But I guess big chunk of the time is spent creating the Kubernetes resources which can be pretty slow from my experience regardless the operator. The more interesting point about the performance test is what happens once the topics are created. E.g. how quickly it shows up when you modify one of them etc.

**Reply by clyang82** [reply]
Thanks @scholzj @fvaleri @see-quick for your information and insights. I would like to share this [architecture](https://github.com/openshift-online/maestro?tab=readme-ov-file#architecture). it takes MQTT as an example. But you can replace MQTT with Kafka. Actually, the transport layer is transparent for us. From this arch, you can see that the transport layer is not a critical component. All the data is stored in DB. And we have re-sync mechanism to guarantee there is no data lost.

These 10k topics are not going to be created all at once. when there is a new maestro agent joining, then create a user and topic. Actually, we need 2 topics per managed cluster.

Since the transport layer is not a critical component, so in the past, I did test with the following configurations in Kafka cluster
```
      default.replication.factor: 1
      inter.broker.protocol.version: "3.5"
      min.insync.replicas: 1
      offsets.topic.replication.factor: 1
      transaction.state.log.min.isr: 1
      transaction.state.log.replication.factor: 1
```
the cpu and memory usage likes
```
NAME                     CPU(cores)   MEMORY(bytes)
my-cluster-kafka-0       33m          8383Mi
my-cluster-kafka-1       43m          6981Mi
my-cluster-kafka-2       37m          6456Mi
my-cluster-zookeeper-0   6m           1555Mi
my-cluster-zookeeper-1   7m           1679Mi
my-cluster-zookeeper-2   10m          1495Mi
```
Please note that I did not use entityoperator. I think create 10K kafkaUser and 10K kafkaTopic CRs is not a good idea in Kubernetes environment. so I called the admin API to achieve it.

**Reply by scholzj** [reply]
MQTT and Kafka are two different things. I do not think you should just replace MQTT with Kafka or the other way around in an architecture like this. The configuration you have also seems to provide no high availability or reliability. While that is possible if you can recover the data from database and assuming your agents can tolerate outages, the general expectation of Kafka is that you will utilize these features. If you aren't, it always raises doubt whether it is the right tool for the task.

**Reply by ppatierno** [reply]
I agree with Jakub. It's not clear but I guess your topics have also just one partition, so you are using Kafka like a kind of queue in the end. Also in your architecture I see the MQTT broker just used a messaging backbone where the server and the agent exchange a message but with no need to re-read the events (something that Kafka supports for example). Also you have a database for storing everything so you are also not relying on the Kafka capability to store data for longer. It sounds that a messaging broker could be the one to use for this scenario (so an MQTT one is fine, but even others using AMQP and so on). Not sure Kafka fits in this picture.
Out of curiosity, what was the rationale behind your decision to use Kafka? Maybe we are missing something.

**Reply by clyang82** [reply]
That always makes sense for me to enable high availability for Kafka. Thanks @scholzj . Based on @see-quick 's test results,  we can create 10K topics with proper configurations for the Kafka cluster. But we do not create 10K topics at once, if we create topic when it needs, what concerns do you have? will it trigger the re-balance so that any topics cannot receive or publish message during the re-balance period?

**Reply by ppatierno** [reply]
I think before talking about HA with Kafka, we need to understand why you picked Kafka and if there are good reasons for that.
Regarding rebalancing ... what are referring to? A rebalancing of multiple partitions to different consumers happens when consumers join/left a consumer group but in your case AFAICS you have one consumer (agent) per topic with just one partition so no rebalance. Also if you are talking about load rebalancing via Cruise Control, it's something you trigger when/if needed, it's not automatic.

**Reply by scholzj** [reply]
> That always makes sense for me to enable high availability for Kafka. Thanks @scholzj . Based on @see-quick 's test results, we can create 10K topics with proper configurations for the Kafka cluster. But we do not create 10K topics at once, if we create topic when it needs, what concerns do you have?

I think there are several aspects you need to understand when it comes to the performance. The initial creations of Xk topics is one part of it. But you should also understand for example when you have 10k existing topics that exist and are ready and you create the 10001st -> will it create in few seconds? Or will it create the new topic in 2 hours? The same applies to when you change some configuration.

The initial creation time is interesting in some usecases. But in general:
* You would rarely create 10k topics at once at the beginning
* Even if you do create 10k topics at once at the beginning, it might not matter that much if it takes one hour

But when you need to create the 10001st topic for a new agent and it takes very long, that will likely matter as something will be waiting for it.

You should also think about the scalability. As already set, Kafka more scaled along partitions than topics. So, having 10k topics with one partition means having 10k partitions. And 10k partitions is not that much for Kafka. But normally 10k partitions will be in a lot less topics. Having 10k topics with single partition is unusual and I think you might run into weird issues and limits because while many users have 10k partitions, a lot of them will have it for example only with 100 topics. Doing unusual things often leads to discovering bugs and issues others don't have.

One has to also wonder where the 10k number comes from. Is it that you in reality expect 100 topics and 10k is the absolutely highest upper limit? Because while 10k might work, 100k or 1M topics will not. So this architecture might not scale well beyond the 10k topics.

> will it trigger the re-balance so that any topics cannot receive or publish message during the re-balance period?

You might have to wait for the next topic to be created or updated. But the existing topics in Kafka in general don't care about the Topic Operator and will work fine.

**Reply by clyang82** [reply]
@ppatierno In order to support it in production, we evaluated AMQ Streams / AMQ Broker / [Mochi-MQTT](https://github.com/mochi-mqtt/server). Due to AMQ Broker cannot support authn and authz with mTLS and we need to build some features (HA/authn/authz) by ourselves for Mochi-MQTT, and the AMQ Streams is production ready and meet our requirements, that is why we choose to use Kafka for production. Any suggestions?

**Reply by yanmxa** [reply]
> I think before talking about HA with Kafka, we need to understand why you picked Kafka and if there are good reasons for that. 

Thanks @ppatierno! It's a good question! IMHO the reason to consider using Kafka here is to take advantage of its authentication and authorization mechanism. Because Kafka's ACL/permission management is based on the topic(if I understand correctly), each cluster will need an exclusive topic for permission control. And a lot of clusters will need to create a lot of topics(10k)...

A simple scenario: clusterA cannot consume the messages for clusterB. Also, clusterA cannot pretend to be clusterB to send messages to server-side by the transport.

> Regarding rebalancing ... what are referring to? A rebalancing of multiple partitions to different consumers happens when consumers join/left a consumer group but in your case AFAICS you have one consumer (agent) per topic with just one partition so no rebalance. Also if you are talking about load rebalancing via Cruise Control, it's something you trigger when/if needed, it's not automatic.

The rebalance @clyang82 referred to is on the server side. where the server subscribes to a pattern topic like `^status.*`. Then the operator finds a new cluster named `clusterX` joined and creates the associated topic `status.clusterX`. Is the rebalance triggered on the server side? I think it happens, cause the new topic/partition will join the server-side consumer group.

**Reply by ppatierno** [reply]
> Due to AMQ Broker cannot support authn and authz with mTLS ...

Is it really true? It sounds really weird to me that upstream Artemis (so AMQ Broker) doesn't have such a feature. @brusdev any thoughts as maintainer on that project?

I am asking because I see it as a scenario for a messaging broker whatever is the protocol. So maybe not using MQTT but i.e. AMQP instead.

**Reply by brusdev** [reply]
@clyang82 @ppatierno I confirm AMQ Broker support authn and authz with mTLS, you can use the [CertificateLoginModule](https://activemq.apache.org/components/artemis/documentation/latest/security.html#certificateloginmodule), for further detail see https://activemq.apache.org/components/artemis/documentation/latest/security.html#dual-authentication

**Reply by ppatierno** [reply]
> Thanks @ppatierno! It's a good question! IMHO the reason to consider using Kafka here is to take advantage of its authentication and authorization mechanism. Because Kafka's ACL/permission management is based on the topic(if I understand correctly), each cluster will need an exclusive topic for permission control. And a lot of clusters will need to create a lot of topics(10k)...
A simple scenario: clusterA cannot consume the messages for clusterB. Also, clusterA cannot pretend to be clusterB to send messages to server-side by the transport.

I think I have got the scenario, so for this reason I am wondering why a messaging broker doesn't work for you. Having maybe one queue per maestro agent? With authn/authz on it? Something where the message flows, is consumed and then deleted because you save everything in a database. While Kafka is providing more features which you don't use.

> Is the rebalance triggered on the server side? I think it happens, cause the new topic/partition will join the server-side consumer group.

Well technically it's not a topic partition joining a consumer group. It's a consumer joining the group. So the server will be the only one member of such a group subscribing to a pattern. Said that I have to verify if a new topic matches the pattern it will trigger a rebalance where anyway the additional partition will be assigned still to the same single consumer.

**Reply by see-quick** [reply]
> I think there are several aspects you need to understand when it comes to the performance. The initial creations of Xk topics is one part of it. But you should also understand for example when you have 10k existing topics that exist and are ready and you create the 10001st -> will it create in few seconds? Or will it create the new topic in 2 hours? The same applies to when you change some configuration.

Just to add my one bit to the performance side...I did not try/test the creation of the 10001st `KafkaTopic` but I tried to update `10 KafkaTopics` configs (already created) and it took around `3s` to complete (for all 10 topics to change a config).

**Comment by clyang82** [comment]
> @clyang82 @ppatierno I confirm AMQ Broker support authn and authz with mTLS, you can use the [CertificateLoginModule](https://activemq.apache.org/components/artemis/documentation/latest/security.html#certificateloginmodule), for further detail see https://activemq.apache.org/components/artemis/documentation/latest/security.html#dual-authentication

Thanks @brusdev for your information. Here is an original issue logged - https://github.com/artemiscloud/activemq-artemis-operator/issues/724. Could you give us the advice? Thanks.

**Reply by brusdev** [reply]
@clyang82 I replied in https://github.com/artemiscloud/activemq-artemis-operator/issues/724. It is already resolved and will be closed when the documentation that explains how to set up the dual authentication by using the ActiveMQArtemis CR is added.

**Reply by clyang82** [reply]
That is a good news. Thanks @brusdev will try it soon.
